File: app/src/main/java/com/gee12/mytetroid/activities/TetroidActivity.java
package com.gee12.mytetroid.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.menu.MenuBuilder;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GestureDetectorCompat;

import com.gee12.mytetroid.R;
import com.gee12.mytetroid.TetroidTask2;
import com.gee12.mytetroid.data.SettingsManager;
import com.gee12.mytetroid.data.StorageManager;
import com.gee12.mytetroid.logs.ILogger;
import com.gee12.mytetroid.logs.LogManager;
import com.gee12.mytetroid.model.TetroidNode;
import com.gee12.mytetroid.utils.FileUtils;
import com.gee12.mytetroid.utils.ViewUtils;
import com.gee12.mytetroid.views.ActivityDoubleTapListener;
import com.gee12.mytetroid.views.Message;

import org.jetbrains.annotations.NotNull;

import java.io.File;

import lib.folderpicker.FolderPicker;

public abstract class TetroidActivity extends AppCompatActivity
        implements View.OnTouchListener, StorageManager.IStorageInitCallback {

    protected GestureDetectorCompat gestureDetector;
    protected Menu mOptionsMenu;
    protected Toolbar mToolbar;
    protected TextView tvTitle;
    protected TextView tvSubtitle;
    protected LinearLayout mLayoutProgress;
    protected TextView mTextViewProgress;
    protected TetroidTask2 mCurTask;
    protected Intent mReceivedIntent;
    protected boolean mIsFullScreen;
    protected boolean mIsOnCreateProcessed;
    protected boolean mIsGUICreated;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutResourceId());

        this.mReceivedIntent = getIntent();

        this.mToolbar = findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);
//        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        setVisibilityActionHome(true);

        // обработчик нажатия на экране
        this.gestureDetector = new GestureDetectorCompat(this,
                new ActivityDoubleTapListener(() -> toggleFullscreen(true)));

        this.tvTitle = mToolbar.findViewById(R.id.text_view_title);
        this.tvSubtitle = mToolbar.findViewById(R.id.text_view_subtitle);

        this.mLayoutProgress = findViewById(R.id.layout_progress_bar);
        this.mTextViewProgress = findViewById(R.id.progress_text);

        this.mIsOnCreateProcessed = false;
        this.mIsGUICreated = false;
    }

    /**
     * Установка пометки, что обработчик OnCreate был вызван, и можно вызвать другие обработчики,
     *  следующие за ним (а не вразнобой на разных устройствах).
     */
    protected void afterOnCreate() {
        this.mIsOnCreateProcessed = true;
        if (mOptionsMenu != null) {
            onCreateOptionsMenu(mOptionsMenu);
            onPrepareOptionsMenu(mOptionsMenu);
        }
    }

    /**
     *
     * @return
     */
    protected abstract int getLayoutResourceId();

    /**
     * Обработчик события, когда создались все элементы интерфейса.
     * Вызывается из onCreateOptionsMenu(), который, в свою очередь, принудительно вызывается после onCreate().
     */
    protected abstract void onGUICreated();

    /**
     * Установка заголовка активности.
     * @param title
     */
    @Override
    public void setTitle(CharSequence title) {
        tvTitle.setText(title);
    }

    /**
     * Установка подзаголовка активности.
     * @param title
     */
    public void setSubtitle(CharSequence title) {
        tvSubtitle.setText(title);
    }

//    /**
//     * Если потеряли фокус на активности, то выходим их полноэкранного режима
//     * (например, при нажатии на "физическую" кнопку вызова меню).
//     * @param hasFocus
//     */
//    @Override
//    public void onWindowFocusChanged(boolean hasFocus) {
//        super.onWindowFocusChanged(hasFocus);
//        if (!hasFocus) {
//            setFullscreen(false);
//        }
//    }

    /**
     * Включение/отключение полноэкранного режима.
     * Оставлен только в RecordActivity.
     * @param fromDoubleTap
     * @return
     */
    public int toggleFullscreen(boolean fromDoubleTap) {
        if (this instanceof RecordActivity) {
            if (!fromDoubleTap || SettingsManager.isDoubleTapFullscreen(this)) {
                boolean newValue = !mIsFullScreen;
                ViewUtils.setFullscreen(this, newValue);
                this.mIsFullScreen = newValue;
                return (newValue) ? 1 : 0;
            }
        } else {
            return -1;
        }
        return -1;
    }

    /**
     * Переопределяем обработчик нажатия на экране
     * для обработки перехода в полноэкранный режим.
     * @param v
     * @param event
     * @return
     */
    @Override
    public boolean onTouch(View v, MotionEvent event) {
        gestureDetector.onTouchEvent(event);
        return false;
    }

    public boolean onBeforeCreateOptionsMenu(Menu menu) {
        boolean onCreateCalled = isOnCreateProcessed();
        if (!onCreateCalled) {
            this.mOptionsMenu = menu;
        }
        return onCreateCalled;
    }

    @SuppressLint("RestrictedApi")
    public boolean onAfterCreateOptionsMenu(Menu menu) {
        // запускаем только 1 раз
        if (!mIsGUICreated) {
            // для отображения иконок
            if (menu instanceof MenuBuilder){
                MenuBuilder m = (MenuBuilder) menu;
                m.setOptionalIconsVisible(true);
            }
            // устанавливаем флаг, что стандартные элементы активности созданы
            onGUICreated();
        }
        this.mIsGUICreated = true;
        return true;
    }

/*    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if (!isOnCreateCalled())
            return true;
        return false;
    }*/

    public void updateOptionsMenu() {
        if (mOptionsMenu != null) {
            onPrepareOptionsMenu(mOptionsMenu);
        } else {
            LogManager.log(this, getString(R.string.log_mOptionsMenu_is_null), ILogger.Types.WARNING, -1);
        }
    }

    /**
     * Обработчик выбора пунктов системного меню.
     * @param item
     * @return
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_fullscreen:
                toggleFullscreen(false);
                return true;
            case R.id.action_about_app:
                ViewUtils.startActivity(this, AboutActivity.class, null);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Обработка возвращаемого результата других активностей.
     *
     * @param requestCode
     * @param resultCode
     * @param data
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if ((requestCode == StorageManager.REQUEST_CODE_OPEN_STORAGE
                || requestCode == StorageManager.REQUEST_CODE_CREATE_STORAGE)
                && resultCode == RESULT_OK) {
            String folderPath = data.getStringExtra(FolderPicker.EXTRA_DATA);
            if (!TextUtils.isEmpty(folderPath)) {
                boolean isCreate = (requestCode == StorageManager.REQUEST_CODE_CREATE_STORAGE);
                openOrCreateStorage(folderPath, isCreate);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NotNull String[] permissions, @NotNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        boolean permGranted = (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED);
        switch (requestCode) {
            case StorageManager.REQUEST_CODE_PERMISSION_WRITE_STORAGE: {
                if (permGranted) {
                    LogManager.log(this, R.string.log_write_ext_storage_perm_granted, ILogger.Types.INFO);
                    loadStorage(null);
                } else {
                    LogManager.log(this, R.string.log_missing_read_ext_storage_permissions, ILogger.Types.WARNING, Toast.LENGTH_SHORT);
                }
            } break;
            case StorageManager.REQUEST_CODE_PERMISSION_TERMUX: {
                if (permGranted) {
                    LogManager.log(this, R.string.log_run_termux_commands_perm_granted, ILogger.Types.INFO);
                    StorageManager.startStorageSyncAndInit(this);
                } else {
                    LogManager.log(this, R.string.log_missing_run_termux_commands_permissions,
                            ILogger.Types.WARNING, Toast.LENGTH_SHORT);
                }
            } break;
        }
    }

    /**
     * Открытие существующего или создание нового хранилище в указанном каталоге.
     * @param folderPath
     * @param isCreate
     */
    private void openOrCreateStorage(String folderPath, boolean isCreate) {
        if (isCreate) {
            if (FileUtils.isDirEmpty(new File(folderPath))) {
                createStorage(folderPath/*, true*/);
            } else {
                LogManager.log(this, R.string.log_dir_not_empty, ILogger.Types.ERROR, Toast.LENGTH_LONG);
            }
        } else {
//            StorageManager.initOrSyncStorage(this, folderPath, true);
            loadStorage(folderPath);
        }
        // сохраняем путь
        SettingsManager.setLastChoosedFolder(this, folderPath);
    }

    /**
     * Старт загрузки существующего хранилища.
     */
    protected abstract void loadStorage(String folderPath);

    /**
     * Старт создания нового хранилища.
     * @param storagePath
     */
    protected abstract void createStorage(String storagePath);

    @Override
    public void blockInterface() {
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
    }

    @Override
    public void unblockInterface() {
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
    }

    @Override
    public void taskStarted(TetroidTask2 task) {
        this.mCurTask = task;
    }

    public boolean isCurTaskRunning() {
        return (mCurTask != null && mCurTask.isRunning());
    }

    @Override
    public int taskPreExecute(int sRes) {
        mTextViewProgress.setText(sRes);
        mLayoutProgress.setVisibility(View.VISIBLE);
        ViewUtils.hideKeyboard(this, getWindow().getDecorView());
        return Gravity.NO_GRAVITY;
    }

    @Override
    public void taskPostExecute(int openedDrawer) {
        mLayoutProgress.setVisibility(View.INVISIBLE);
    }

    @Override
    public void initGUI(boolean res, boolean mIsFavoritesOnly, boolean mIsOpenLastNode) {

    }

    @Override
    public void afterStorageLoaded(boolean res) {

    }

    @Override
    public void afterStorageDecrypted(TetroidNode node) {

    }

    /**
     * Обработчик, вызываемый перед запуском кода в обработчике onBackPressed().
     * @return true - можно продолжить работу обработчика onBackPressed(), иначе - прервать
     */
    public boolean onBeforeBackPressed() {
        if (isCurTaskRunning()) {
            // если выполняется задание, то не реагируем на нажатие кнопки Back
            return false;
        }
        return true;
    }

    protected void setVisibilityActionHome(boolean isVis) {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(isVis);
        }
    }

    /**
     * Установка видимости пункта меню.
     * @param menuItem
     * @param isVisible
     */
    protected void visibleMenuItem(MenuItem menuItem, boolean isVisible) {
        ViewUtils.setVisibleIfNotNull(menuItem, isVisible);
    }

    /**
     * Установка активности пункта меню.
     * @param menuItem
     * @param isEnabled
     */
    protected void enableMenuItem(MenuItem menuItem, boolean isEnabled) {
        ViewUtils.setEnabledIfNotNull(menuItem, isEnabled);
    }

    /**
     * Установка подзаголовка активности.
     * @param subtitle
     */
    protected void setSubtitle(String subtitle) {
        tvSubtitle.setVisibility(View.VISIBLE);
        tvSubtitle.setTextSize(16);
        tvSubtitle.setText(subtitle);
    }

    public void showActivityForResult(Class<?> cls, int requestCode) {
        Intent intent = new Intent(this, cls);
        startActivityForResult(intent, requestCode);
    }

    public boolean isOnCreateProcessed() {
        return mIsOnCreateProcessed;
    }

    /**
     * Вывод интерактивного уведомления SnackBar "Подробнее в логах".
     */
    protected void showSnackMoreInLogs() {
        Message.showSnackMoreInLogs(this, R.id.layout_coordinator);
    }
}

File: app/src/main/java/com/gee12/mytetroid/ui/base/TetroidActivity.kt
package com.gee12.mytetroid.ui.base

import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.*
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.menu.MenuBuilder
import androidx.appcompat.view.menu.MenuPopupHelper
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GestureDetectorCompat
import androidx.core.view.isVisible
import androidx.documentfile.provider.DocumentFile
import androidx.lifecycle.lifecycleScope
import com.anggrayudi.storage.SimpleStorageHelper
import com.anggrayudi.storage.file.FileFullPath
import com.gee12.mytetroid.R
import com.gee12.mytetroid.common.extensions.hideKeyboard
import com.gee12.mytetroid.common.utils.ViewUtils
import com.gee12.mytetroid.data.settings.CommonSettings
import com.gee12.mytetroid.di.ScopeSource
import com.gee12.mytetroid.domain.manager.CommonSettingsManager
import com.gee12.mytetroid.domain.provider.BuildInfoProvider
import com.gee12.mytetroid.domain.provider.IAppPathProvider
import com.gee12.mytetroid.domain.provider.IResourcesProvider
import com.gee12.mytetroid.logs.ITetroidLogger
import com.gee12.mytetroid.logs.LogType
import com.gee12.mytetroid.logs.Message
import com.gee12.mytetroid.model.permission.PermissionRequestCode
import com.gee12.mytetroid.model.permission.TetroidPermission
import com.gee12.mytetroid.ui.TetroidMessage
import com.gee12.mytetroid.ui.about.AboutAppActivity
import com.gee12.mytetroid.ui.base.views.ActivityDoubleTapListener
import com.gee12.mytetroid.ui.record.RecordActivity
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.koin.core.scope.Scope
import org.koin.core.scope.get

abstract class TetroidActivity<VM : BaseViewModel>
    : AppCompatActivity(), ITetroidComponent, View.OnTouchListener {

    interface IDownloadFileResult {
        fun onSuccess(uri: Uri)
        fun onError(ex: Exception)
    }

    protected lateinit var scopeSource: ScopeSource
    protected val koinScope: Scope
        get() = scopeSource.scope

    lateinit var viewModel: VM

    val resourcesProvider: IResourcesProvider by inject()
    val settingsManager: CommonSettingsManager by inject()
    val appPathProvider: IAppPathProvider by inject()
    val buildInfoProvider: BuildInfoProvider by inject()
    val logger: ITetroidLogger by inject()

    protected var receivedIntent: Intent? = null
    var optionsMenu: Menu? = null
    // toolbar
    protected var toolbar: Toolbar? = null
    protected var tvSingleTitle: TextView? = null
    protected var tvTitle: TextView? = null
    protected var tvSubtitle: TextView? = null
    // progress bar
    protected var layoutProgress: LinearLayout? = null
    protected var tvProgress: TextView? = null

    protected lateinit var gestureDetector: GestureDetectorCompat
    protected var isFullScreen = false
    protected var isOnCreateProcessed = false
    protected var isUiCreated = false

    var fileStorageHelper: SimpleStorageHelper? = null

    private var requestCodeForStorageAccess: PermissionRequestCode? = null

    // region Create

    protected abstract fun getLayoutResourceId(): Int

    protected abstract fun getViewModelClazz(): Class<VM>

    protected open fun hasToolbar() = true

    protected open fun hasProgressBar() = true

    protected open fun isSingleTitle() = true

    protected open fun isAppearInLogs() = true


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutResourceId())

        receivedIntent = intent

        if (hasToolbar()) {
            toolbar = findViewById(R.id.toolbar)
            setSupportActionBar(toolbar)
            setVisibilityActionHome(true)

            tvSingleTitle = toolbar?.findViewById(R.id.text_view_single_title)
            tvTitle = toolbar?.findViewById(R.id.text_view_title)
            tvSubtitle = toolbar?.findViewById(R.id.text_view_subtitle)

            setTitle(title)
        }

        if (hasProgressBar()) {
            layoutProgress = findViewById(R.id.layout_progress_bar)
            tvProgress = findViewById(R.id.progress_text)
        }

        // обработчик нажатия на экране
        gestureDetector = GestureDetectorCompat(this,
            ActivityDoubleTapListener {
                toggleFullscreen(true)
            })
        isOnCreateProcessed = false
        isUiCreated = false

        if (isUseFileStorage()) {
            fileStorageHelper = SimpleStorageHelper(this, savedInstanceState)
            // ради оптимизации удаляются "повторные" persistable uri, к которым юзер предоставил доступ в SAF,
            // и оставляются только "родительские", из-за чего происходит циклический запрос доступа на Android >= 11
            fileStorageHelper?.storage?.isCleanupRedundantUriPermissions = false
            initFileStorage(savedInstanceState)
        }

        createDependencyScope()
        afterCreateDependencyScope()
        createViewModel()
        initViewModel()
    }

    protected open fun createDependencyScope() {
        scopeSource = ScopeSource.createNew()
    }

    protected open fun afterCreateDependencyScope() {
    }

    protected open fun createViewModel() {
        viewModel = koinScope.get(getViewModelClazz())
    }

    protected open fun initViewModel() {
        viewModel.initialize()

        if (isAppearInLogs()) {
            logger.logDebug(getString(R.string.log_activity_opened_mask, javaClass.simpleName))
        }

        lifecycleScope.launch {
            viewModel.eventFlow.collect { event -> onBaseEvent(event) }
        }
        lifecycleScope.launch {
            viewModel.messageEventFlow.collect { message -> showMessage(message) }
        }
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onStart() {
        super.onStart()
        viewModel.setNotificatorCallbacks()
    }

    /**
     * Установка пометки, что обработчик OnCreate был вызван, и можно вызвать другие обработчики,
     * следующие за ним (а не вразнобой на разных устройствах).
     */
    protected fun afterOnCreate() {
        isOnCreateProcessed = true
        optionsMenu?.let {
            onCreateOptionsMenu(it)
            onPrepareOptionsMenu(it)
        }
    }

    /**
     * Обработчик события, когда создались все элементы интерфейса.
     * Вызывается из onCreateOptionsMenu(), который, в свою очередь, принудительно вызывается после onCreate().
     */
    protected open fun onUiCreated() {
    }

    // endregion Create

    // region Events

    /**
     * Обработчик изменения состояния View.
     */
    protected open fun onBaseEvent(event: BaseEvent) {
        when (event) {
            is BaseEvent.ShowProgress -> setProgressVisibility(true)
            is BaseEvent.HideProgress -> setProgressVisibility(false)
            is BaseEvent.ShowProgressWithText -> showProgress(event.message)
            is BaseEvent.TaskStarted -> {
                taskPreExecute(event.titleResId ?: R.string.state_loading)
            }
            BaseEvent.TaskFinished -> taskPostExecute()
            BaseEvent.ShowMoreInLogs -> showSnackMoreInLogs()
            else -> Unit
        }
    }

    // endregion Events

    // region Permissions

    protected open fun onPermissionGranted(
        permission: TetroidPermission?,
        requestCode: PermissionRequestCode,
    ) = Unit

    protected open fun onPermissionCanceled(
        requestCode: PermissionRequestCode
    ) = Unit

    // endregion Permissions

    // region File

    override fun isUseFileStorage() = false

    protected fun requestFileStorageAccess(
        uri: Uri,
        requestCode: PermissionRequestCode,
    ) {
        requestCodeForStorageAccess = requestCode
        fileStorageHelper?.requestStorageAccess(
        //fileStorageHelper.openFolderPicker(
            requestCode = requestCode.code,
            initialPath = FileFullPath(this, fullPath = uri.toString()),
        )
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        fileStorageHelper?.storage?.checkIfFileReceived(intent)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        fileStorageHelper?.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        fileStorageHelper?.onRestoreInstanceState(savedInstanceState)
    }

    // TODO: в отдельный FileStorageManager
    private fun initFileStorage(savedInstanceState: Bundle?) {
        savedInstanceState?.let { fileStorageHelper?.onRestoreInstanceState(it) }
        fileStorageHelper?.onStorageAccessGranted = { requestCode, root ->
            onStorageAccessGranted(requestCode, root)
        }
        fileStorageHelper?.onFileSelected = { requestCode, files ->
            onFileSelected(requestCode, files)
        }
        fileStorageHelper?.onFolderSelected = { requestCode, folder ->
            // сохраняем путь
            settingsManager.setLastSelectedFolderPath(path = folder.uri.toString())

            if (requestCodeForStorageAccess?.code == requestCode) {
                onStorageAccessGranted(requestCode, folder)
            } else {
                onFolderSelected(requestCode, folder)
            }
        }
        /*fileStorageHelper.onFileCreated = { requestCode, file ->
            onFileCreated(requestCode, file)
        }
        fileStorageHelper.onFileReceived = object : SimpleStorageHelper.OnFileReceived {
            override fun onFileReceived(files: List<DocumentFile>) {
                val names = files.joinToString(", ") { it.fullName }
                Toast.makeText(baseContext, "File received: $names", Toast.LENGTH_SHORT).show()
            }

            override fun onNonFileReceived(intent: Intent) {
                Toast.makeText(baseContext, "Non-file is received", Toast.LENGTH_SHORT).show()
            }
        }*/
        if (savedInstanceState == null) {
            fileStorageHelper?.storage?.checkIfFileReceived(intent)
        }
    }

    override fun onStorageAccessGranted(requestCode: Int, root: DocumentFile) {}

    override fun onFolderSelected(requestCode: Int, folder: DocumentFile) {}

    override fun onFileSelected(requestCode: Int, files: List<DocumentFile>) {}

    //override fun onFileCreated(requestCode: Int, file: DocumentFile) {}

    // TODO: в отдельный FileStorageManager или TetroidActivityComponent (для android-зависимой логики)
    fun openFilePicker(
        requestCode: PermissionRequestCode,
        allowMultiple: Boolean = false,
        initialPath: String? = null,
        filterMimeTypes: Array<String> = emptyArray(),
    ) {
        val path = initialPath ?: settingsManager.getLastSelectedFolderPathOrDefault(true)
        try {
            fileStorageHelper?.openFilePicker(
                requestCode = requestCode.code,
                allowMultiple = allowMultiple,
                initialPath = path?.let { FileFullPath(this, fullPath = it) },
                filterMimeTypes = filterMimeTypes,
            )
        } catch (ex: Exception) {
            try {
                fileStorageHelper?.openFilePicker(
                    requestCode = requestCode.code,
                    allowMultiple = allowMultiple,
                    initialPath = null,
                    filterMimeTypes = filterMimeTypes,
                )
                viewModel.logWarning(getString(R.string.error_incorrent_initial_path_for_picker_mask, initialPath), show = false)
            } catch (ex: Exception) {
                viewModel.logError(getString(R.string.error_open_file_folder_picker), show = true)
            }
        }
    }

    // TODO: в отдельный FileStorageManager
    fun openFolderPicker(
        requestCode: PermissionRequestCode,
        initialPath: String? = null,
    ) {
        val path = initialPath ?: settingsManager.getLastSelectedFolderPathOrDefault(true)
        try {
            fileStorageHelper?.openFolderPicker(
                requestCode = requestCode.code,
                initialPath = path?.let { FileFullPath(this, fullPath = it) },
            )
        } catch (ex: Exception) {
            try {
                fileStorageHelper?.openFolderPicker(
                    requestCode = requestCode.code,
                    initialPath = null,
                )
                viewModel.logWarning(getString(R.string.error_incorrent_initial_path_for_picker_mask, initialPath), show = false)
            } catch (ex: Exception) {
                viewModel.logError(getString(R.string.error_open_file_folder_picker), show = true)
            }
        }
    }

    // endregion File

    /**
     * Установка заголовка активности.
     */
    override fun setTitle(title: CharSequence?) {
        val isSingleTitle = isSingleTitle()
        tvTitle?.isVisible = !isSingleTitle
        tvSubtitle?.isVisible = !isSingleTitle
        tvSingleTitle?.isVisible = isSingleTitle
        if (isSingleTitle) {
            tvSingleTitle?.text = title
        } else {
            tvTitle?.text = title
        }
    }

    /**
     * Установка подзаголовка активности.
     */
    fun setSubtitle(title: CharSequence?) {
        tvSubtitle?.text = title
    }

    //    /**
    //     * Если потеряли фокус на активности, то выходим их полноэкранного режима
    //     * (например, при нажатии на "физическую" кнопку вызова меню).
    //     * @param hasFocus
    //     */
    //    @Override
    //    public void onWindowFocusChanged(boolean hasFocus) {
    //        super.onWindowFocusChanged(hasFocus);
    //        if (!hasFocus) {
    //            setFullscreen(false);
    //        }
    //    }

    /**
     * Включение/отключение полноэкранного режима.
     * Оставлен только в RecordActivity.
     */
    open fun toggleFullscreen(fromDoubleTap: Boolean): Int {
        if (this is RecordActivity) {
            if (!fromDoubleTap || CommonSettings.isDoubleTapFullscreen(this)) {
                val newValue = !isFullScreen
                ViewUtils.setFullscreen(this, newValue)
                isFullScreen = newValue
                return if (newValue) 1 else 0
            }
        } else {
            return -1
        }
        return -1
    }

    /**
     * Переопределяем обработчик нажатия на экране
     * для обработки перехода в полноэкранный режим.
     */
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        gestureDetector.onTouchEvent(event)
        return false
    }

    fun onBeforeCreateOptionsMenu(menu: Menu?): Boolean {
        val onCreateCalled = isOnCreateProcessed
        if (!onCreateCalled) {
            optionsMenu = menu
        }
        return onCreateCalled
    }

    @SuppressLint("RestrictedApi")
    fun onAfterCreateOptionsMenu(menu: Menu?): Boolean {
        // запускаем только 1 раз
        if (!isUiCreated) {
            // для отображения иконок
            if (menu is MenuBuilder) {
                menu.setOptionalIconsVisible(true)
            }
        }
        // устанавливаем флаг, что стандартные элементы активности созданы
        onUiCreated()
        isUiCreated = true
        return true
    }

    fun updateOptionsMenu() {
        optionsMenu?.let {
            onPrepareOptionsMenu(it)
        }
    }

    /**
     * Обработчик выбора пунктов системного меню.
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_fullscreen -> {
                toggleFullscreen(false)
                return true
            }
            R.id.action_about_app -> {
                AboutAppActivity.start(this)
                return true
            }
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Обработчик результата активити.
     */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        // TODO: ?
        // разрешение на запись в память на Android 11 запрашивается с помощью Intent
        if (requestCode == PermissionRequestCode.OPEN_STORAGE_FOLDER.code) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val isGranted = (viewModel.permissionManager.hasWriteExtStoragePermission(this))
                onRequestPermissionsResult(
                    requestCode = requestCode,
                    permissions = emptyArray(),
                    grantResults = IntArray(1) {
                        if (isGranted) PackageManager.PERMISSION_GRANTED else PackageManager.PERMISSION_DENIED }
                )
            }
        }
    }

    /**
     * Обработчик запроса разрешения.
     */
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        val isGranted = grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
        val permissionRequestCode = PermissionRequestCode.fromCode(requestCode)
        permissionRequestCode?.also {
            if (isGranted) {
                viewModel.log(getString(R.string.log_permission_granted_mask, it.name))
                onPermissionGranted(permission = null, requestCode = it)
            } else {
                viewModel.log(getString(R.string.log_permission_canceled_mask, it.name))
                onPermissionCanceled(requestCode = it)
            }
        }
    }

    fun taskPreExecute(progressTextResId: Int) {
        blockInterface()
        showProgress(progressTextResId)
        window.decorView.hideKeyboard()
    }

    fun taskPostExecute() {
        unblockInterface()
        setProgressVisibility(isVisible = false)
    }

    fun blockInterface() {
        window.setFlags(
            WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
            WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
        )
    }

    fun unblockInterface() {
        window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
    }

    /**
     * Обработчик, вызываемый перед запуском кода в обработчике onBackPressed().
     * @return true - можно продолжить работу обработчика onBackPressed(), иначе - прервать
     */
    fun onBeforeBackPressed(): Boolean {
        // если выполняется задание, то не реагируем на нажатие кнопки Back
        return !viewModel.isBusy
    }

    protected fun setVisibilityActionHome(isVisible: Boolean) {
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(isVisible)
    }

    // region IAndroidComponent

    override fun setProgressVisibility(isVisible: Boolean, text: String?) {
        layoutProgress?.isVisible = isVisible
        tvProgress?.text = text
    }

    override fun showProgress(textResId: Int) {
        showProgress(text = getString(textResId))
    }

    override fun showProgress(text: String?) {
        setProgressVisibility(isVisible = true, text)
    }

    override fun hideProgress() {
        setProgressVisibility(isVisible = false)
    }

    /**
     * Публикация сообщений.
     */
    fun showMessage(message: Message) {
        TetroidMessage.show(this, message)
    }

    protected fun showMessage(messageResId: Int, type: LogType = LogType.INFO) {
        showMessage(getString(messageResId), type)
    }

    protected fun showMessage(message: String, type: LogType = LogType.INFO) {
        showMessage(Message(message, type))
    }

    // endregion IAndroidComponent

    /**
     * Вывод интерактивного уведомления SnackBar "Подробнее в логах".
     */
    override fun showSnackMoreInLogs() {
        TetroidMessage.showSnackMoreInLogs(this, R.id.layout_coordinator)
    }
    /**
     * Установка видимости пункта меню.
     */
    protected fun visibleMenuItem(menuItem: MenuItem?, isVisible: Boolean) {
        ViewUtils.setVisibleIfNotNull(menuItem, isVisible)
    }

    /**
     * Установка активности пункта меню.
     */
    protected fun enableMenuItem(menuItem: MenuItem?, isEnabled: Boolean) {
        ViewUtils.setEnabledIfNotNull(menuItem, isEnabled)
    }

    /**
     * Принудительное отображение иконок у пунктов меню.
     */
    @SuppressLint("RestrictedApi")
    protected fun setForceShowMenuIcons(v: View, menu: MenuBuilder) {
        val menuHelper = MenuPopupHelper(this, menu, v)
        menuHelper.setForceShowIcon(true)
        menuHelper.show()
    }

    /**
     * Установка подзаголовка активности.
     */
    fun setSubtitle(subtitle: String?) {
        tvSubtitle?.visibility = View.VISIBLE
        tvSubtitle?.textSize = 16f
        tvSubtitle?.text = subtitle
    }

    open fun showActivityForResult(cls: Class<*>?, requestCode: Int) {
        val intent = Intent(this, cls)
        startActivityForResult(intent, requestCode)
    }

}